doc/man3/X509_cmp.pod

```diff
@@ -0,0 +1,79 @@
+=pod
+
+=head1 NAME
+
+X509_cmp, X509_NAME_cmp,
+X509_issuer_and_serial_cmp, X509_issuer_name_cmp, X509_subject_name_cmp,
+X509_CRL_cmp, X509_CRL_match
+- compare X509 certificates and related values
+
+=head1 SYNOPSIS
+
+ #include <openssl/x509.h>
+
+ int X509_cmp(const X509 *a, const X509 *b);
+ int X509_NAME_cmp(const X509_NAME *a, const X509_NAME *b);
+ int X509_issuer_and_serial_cmp(const X509 *a, const X509 *b);
+ int X509_issuer_name_cmp(const X509 *a, const X509 *b);
+ int X509_subject_name_cmp(const X509 *a, const X509 *b);
+ int X509_CRL_cmp(const X509_CRL *a, const X509_CRL *b);
+ int X509_CRL_match(const X509_CRL *a, const X509_CRL *b);
+
+=head1 DESCRIPTION
+
+This set of functions are used to compare X509 objects, including X509
+certificates, X509 CRL objects and various values in an X509 certificate.
+
+The X509_cmp() function compares two B<X509> objects indicated by parameters
+B<a> and B<b>. The comparison is based on the B<memcmp> result of the SHA-1
+hash values of two B<X509> objects and the canonical (DER) encoding values.
+
+The X509_NAME_cmp() function compares two B<X509_NAME> objects indicated by
+parameters B<a> and B<b>. The comparison is based on the B<memcmp> result of
+the canonical (DER) encoding values of the two objects. L<i2d_X509_NAME(3)>
+has a more detailed description on the DER encoding of B<X509_NAME> structure.
+
+The X509_issuer_and_serial_cmp() function compares the serial number and issuer
+values in the given B<X509> objects B<a> and B<b>.
+
+The X509_issuer_name_cmp(), X509_subject_name_cmp() and X509_CRL_cmp() functions
+are effectively wrappers of X509_NAME_cmp() function. These functions compare 
+issuer names and subject names of the X<509> objects, or issuers of B<X509_CRL>
+objects, respectively.
+
+The X509_CRL_match() function compares two B<X509_CRL> objects. Unlike the
+X509_CRL_cmp() function, this function compares the SHA-1 hash values of the
+whole CRL content indicated by parameter B<a> and B<b>.
+
+=head1 RETURN VALUES
+
+Like common memory comparison functions, the B<X509> comparison functions return
+an integer less than, equal to, or greater than zero if object B<a> is found,
+respectively, to be less than, to match, or be greater than object B<b>.
+
+X509_NAME_cmp(), X509_issuer_and_serial_cmp(), X509_issuer_name_cmp(),
+X509_subject_name_cmp() and X509_CRL_cmp() may return B<-2> to indicate an error.
+
+=head1 NOTES
+
+These functions in fact utilize the underlying B<memcmp> of C library to do the
+comparison job. Data to be compared varies from DER encoding data, SHA-1 hash
+value or B<ASN1_STRING>.
+
+X509_NAME_cmp() and wrappers utilize B<-2> value to indicated errors in some
+circumstances, which could cause confusion for the applications.
```

Comment:
Alternative formulation (only a suggestion):
```
X509_NAME_cmp() and wrappers utilize B<-2> value to indicated errors in some
circumstances, so the error value needs to be treated separately before testing
for a nonzero return value.
```

Reply:
This is very hard to describe. From my understandings, `X509_NAME_cmp()` can also return `-2` to indicate that 'a' is 'less' than 'b', so checking error value in advance keeps the confusion. Should this kind description be placed in the `BUGS` section? This does sound like a bug...

Reply:
> This is very hard to describe. ... so checking error value in advance keeps the confusion.

My suggested formulation was simply based on your description, not on source code analysis. That might explain why it does not describe the current situation correctly.

I think it would be very helpful if you could find out and describe the possible error cases in more detail and give the user guidance how to deal with the quirks of the API properly.

>  Should this kind description be placed in the `BUGS` section?

I wouldn't call this a BUG, it's rather suboptimal API design. So IMHO the `NOTES` section is a good place to do it and adding a separate `BUGS` section might only complicate things.